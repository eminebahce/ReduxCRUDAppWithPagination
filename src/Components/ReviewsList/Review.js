import React from 'react'
import {
    Button,
    ListItem,
    ListItemAvatar,
    ListItemText,
    Avatar,
    Typography,
    IconButton,
    Grid,
    Tooltip
} from '@material-ui/core'
import Divider from '@material-ui/core/Divider';
import Delete from '@material-ui/icons/Delete'
import StarRatingComponent from 'react-star-rating-component';
import {deleteReview, loadReviews} from '../../Actions/ReviewActions';
import {connect} from "react-redux";
import PropTypes from 'prop-types';

class Review extends React.Component {

    onDelete = (reviewId) => {
        if (this.props.reviewSize !== 1 || this.props.totalCount === 0) {
            this.props.deleteReview(reviewId, this.calculateCurrentPage());
        } else {
            this.props.deleteReview(reviewId, this.calculatePreviousPage());
        }
    }

    calculateCurrentPage = () => {
        const firstSplit = this.props.currentUrl.split("&");
        const secondSplit = firstSplit[0].split("=");
        return parseInt(secondSplit[1]);
    }

    calculatePreviousPage = () => {
        return this.calculateCurrentPage() - 1;
    }

    handleClick = (url) => {
        this.props.loadReviews(url);
    }

    render() {
        return (
            <div data-test="review">
                {!this.props.reviews && <p>Loading...</p>}
                <div>
                    {this.props.reviews && this.props.reviews.map((review, index) =>
                        <div key={index}>
                            <ListItem alignItems="flex-start">
                                <ListItemAvatar>
                                    <Avatar alt="avatar"
                                            src="/outline-account_circle-black-18/2x/outline_account_circle_black_18dp.png"/>
                                </ListItemAvatar>
                                <ListItemText
                                    primary={
                                        <React.Fragment>
                                            <StarRatingComponent
                                                name="star-rating"
                                                value={review.rating}
                                                editing={false}
                                            />
                                        </React.Fragment>
                                    }
                                    secondary={
                                        <React.Fragment>
                                                <Typography component="span" style={{display: 'inline'}}
                                                            color="textPrimary">
                                                    {`Anonymous User`}
                                                </Typography>
                                            {` - ${review.body}`}
                                        </React.Fragment>
                                    }
                                />
                                <Tooltip title="Click to remove the review">
                                    <IconButton color='primary' onClick={() => this.onDelete(review.id)}>
                                        <Delete/>
                                    </IconButton>
                                </Tooltip>
                            </ListItem>
                            <Divider/>
                        </div>
                    )}
                </div>
                <div>
                    <Grid container
                          direction="row"
                          justify="center"
                          alignItems="center">
                        {this.props.links && Object.keys(this.props.links).map((link, index) =>
                            <Grid item key={index}
                                  role="group">
                                <Button type="button"
                                        color="primary"
                                        variant="outlined"
                                        size="small"
                                        style={{marginRight: '5px', marginTop: '10px'}}
                                        onClick={() => this.handleClick(this.props.links[link].url)}>
                                    {this.props.links[link].rel}
                                </Button>
                            </Grid>
                        )}
                    </Grid>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    reviewSize: state.reviewSize,
    totalCount: state.totalCount,
    currentUrl: state.currentUrl
})

Review.propTypes = {
    onDelete: PropTypes.func,
    calculateCurrentPage: PropTypes.func,
    calculatePreviousPage: PropTypes.func,
    handleClick: PropTypes.func
}

export default connect(mapStateToProps, {deleteReview, loadReviews})(Review);