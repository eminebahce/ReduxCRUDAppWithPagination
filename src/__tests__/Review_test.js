import React from 'react';
import Review from '../Components/ReviewsList/Review';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import store from '../store';
import { shallow } from 'enzyme';
import checkPropTypes from "check-prop-types";
import PropTypes from "prop-types";

describe('<Review /> component', () => {
    beforeEach(() => {
        const props = {
            onDelete: () => {},
            calculateCurrentPage: () => {},
            calculatePreviousPage: () => {},
            handleClick: () => {}
        };
    });
    const checkProps = (component, expectedProps) => {
        const propsErr = checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
        return propsErr;
    };
    describe('Checking propTypes', () => {
        it('should not throw a warning', () => {
            const expectedProps = {
                onDelete: () => {},
                calculateCurrentPage: () => {},
                calculatePreviousPage: () => {},
                handleClick: () => {}
            };
            const propsError = checkProps(Review, expectedProps);
            expect(propsError).toBeUndefined();
        });
    });
    it('should render without error', () => {
        const component = shallow(<Review />);
        expect(component.find(`[data-test='review']`)).toBeTruthy();
    });
    it("should return reviews", () => {
        const component = renderer.create(
            <Provider store={store}>
            <Review/>
            </Provider>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});