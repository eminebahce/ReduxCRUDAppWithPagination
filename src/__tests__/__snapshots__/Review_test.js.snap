// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Review /> component should return reviews 1`] = `
<div
  data-test="review"
>
  <p>
    Loading...
  </p>
  <div />
  <div>
    <div
      className="MuiGrid-container-1 MuiGrid-align-items-xs-center-9 MuiGrid-justify-xs-center-18"
    />
  </div>
</div>
`;
